internal/oci/notation.go

```diff
@@ -0,0 +1,241 @@
+func NewNotaryVerifier(opts ...Options) (*NotaryVerifier, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	store := &trustStore{
+		cert: o.PublicKey,
+	}
+
+	verifier, err := verifier.New(o.TrustStore, store, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case o.Auth != nil && o.Keychain != nil:
+		return nil, fmt.Errorf("please provide either authn.Authenticator or authn.Keychain")
+	case o.Auth == nil:
+		o.Auth = authn.Anonymous
+	}
+
+	return &NotaryVerifier{
+		auth:     o.Auth,
+		keychain: o.Keychain,
+		verifier: &verifier,
+		opts:     o.ROpt,
+		insecure: o.Insecure,
+	}, nil
+}
+
+// Verify verifies the authenticity of the given ref OCI image.
+// It returns a boolean indicating if the verification was successful.
+// It returns an error if the verification fails, nil otherwise.
+func (v *NotaryVerifier) Verify(ctx context.Context, ref name.Reference) (bool, error) {
+	url := ref.Name()
+	remoteRepo, err := oras.NewRepository(url)
+	if err != nil {
+		return false, err
+	}
+	remoteRepo.PlainHTTP = v.insecure
+
+	repo := registry.NewRepository(remoteRepo)
+
+	credentialProvider := func(ctx context.Context, registry string) (oauth.Credential, error) {
+		return oauth.EmptyCredential, nil
+	}
+
+	auth := v.auth
+	if v.keychain != nil {
+		source := stringResource{url}
+
+		auth, err = v.keychain.Resolve(source)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	if auth != authn.Anonymous {
+		authConfig, err := auth.Authorization()
+		if err != nil {
+			return false, err
+		}
+
+		credentialProvider = func(ctx context.Context, registry string) (oauth.Credential, error) {
+			if authConfig.Username != "" || authConfig.Password != "" || authConfig.IdentityToken != "" {
+				return oauth.Credential{
+					Username:     authConfig.Username,
+					Password:     authConfig.Password,
+					RefreshToken: authConfig.IdentityToken,
+				}, nil
+			}
+			return oauth.EmptyCredential, nil
+		}
+	}
+
+	repoClient := &oauth.Client{
+		Header: http.Header{
+			"User-Agent": {"flux"},
+		},
+		Cache:      oauth.NewCache(),
+		Credential: credentialProvider,
+	}
+
+	remoteRepo.Client = repoClient
+
+	i, err := remote.Image(ref, v.opts...)
+	if err != nil {
+		return false, err
+	}
+
+	d, err := i.Digest()
+	if err != nil {
+		return false, err
+	}
+
+	repoUrl := ""
+
+	lastIndex := strings.LastIndex(url, ":")
+	firstPart := url[:lastIndex]
+
+	if s := strings.Split(url, ":"); len(s) >= 2 && !strings.Contains(url, "@") {
+		repoUrl = fmt.Sprintf("%s@%s", firstPart, d)
+	}
+
+	verififyOptions := notation.VerifyOptions{
+		ArtifactReference:    repoUrl,
+		MaxSignatureAttempts: 50,
+	}
+
+	_, signatures, err := notation.Verify(ctx, *v.verifier, repo, verififyOptions)
+	if err != nil {
+		return false, err
+	}
+
+	if len(signatures) == 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// stringResource represents a resource with a string value.
+type stringResource struct {
```

Comment:
Doing so would create an import cycle as `internal/helm/registry` imports `internal/oci` already. Since the registry package already utilises the oci package, perhaps `internal/helm/registry` could reuse the oci 's `stringResource` func?